### PR TITLE
TS-4075: add a state check for sslHandshakeHookState

### PR DESCRIPTION
**Pull request by oknet**
Add a state check for sslHandshakeHookState after PreAcceptHookState checking in sslServerHandShakeEvent().
and modify the codes in reenable() and callHooks() to fit the patch

The Processing:
path A for normal handshake.
path B for ssl session reuse

1. client initial a tcp connection with ATS
2. ATS trigger a PreAccept Hooks
3. PreAccept Hooks Done
4a. client send a "Client Hello with Sever Cert Request"
5a. set handshakestate to CERT from PRE
6a. SSLAccept() got a "Server Cert Request" then trigger callHooks()
7a. set curHooks
8a. if curHook != NULL then set handshakestate to INVOKE and invoke hooks.
9a. reenable in Hooks A
10a. invoke Hook B and next Hooks ... until curHook == NULL
11a. set handshakestate to DONE
12. SSLAccept() handshake finished

4b. client send a "ssl session reuse request"
5b. set handshakestate to CERT from PRE
6b. SSLAccept() got a "ssl session reuse reques" then reuse session handshake finished
7b. set handshakestate to DONE from CERT

### CODE REVIEW

File: iocore/net/SSLNetVConnection.cc [hunk 1]
```diff
@@ -1400,10 +1416,9 @@ SSLNetVConnection::select_next_protocol(SSL *ssl, const unsigned char **out, uns
...
 SSLNetVConnection::reenable(NetHandler *nh)
 {
-  if (this->sslPreAcceptHookState != SSL_HOOKS_DONE) {
-    this->sslPreAcceptHookState = SSL_HOOKS_INVOKE;
-    this->readReschedule(nh);
-  } else {
+  if (sslPreAcceptHookState != SSL_HOOKS_DONE) {
+    sslPreAcceptHookState = SSL_HOOKS_INVOKE;
+  } else if (sslHandshakeHookState == HANDSHAKE_HOOKS_INVOKE) {
     // Reenabling from the handshake callback
```
Comment (shinrich) [inline iocore/net/SSLNetVConnection.cc]: I'll try to play with this patch over break.  As I recall the readReschedule was necessary to deal with multi-thread processing.  But I could be mis-remembering.  Or cleaning up the state transitions may have eliminated that problem.

**Comment by shinrich** [comment]
Overall looks good.  I want to work with the patch a bit before committing to try out a few different callback scenarios.

File: iocore/net/SSLNetVConnection.cc (hunk 1 above)
Comment (oknet) [inline iocore/net/SSLNetVConnection.cc]: I put the "this->readReschedule(nh)" at the end of reenable.

File: iocore/net/SSLNetVConnection.cc [hunk 2]
```diff
@@ -1095,6 +1105,12 @@ SSLNetVConnection::sslServerHandShakeEvent(int &err)
   }
 
   ssl_error_t ssl_error = SSLAccept(ssl);
+  // if non SNI-Hook set before or SSL Session reuse here:
+  //   the sslHandshakeHookState should be HOOKS_CERT after SSLAccept().
+  // thus, set it to HOOKS_DONE directly.
+  if (HANDSHAKE_HOOKS_CERT == sslHandshakeHookState) {
+    sslHandshakeHookState = HANDSHAKE_HOOKS_DONE;
+  }
```
Comment (oknet) [inline iocore/net/SSLNetVConnection.cc]: please hold this pull request，I found a bug on it，and will release a new patch.
set HookState to DONE if SSLAccept() only send a Server Hello and without certificate.
next time, SSLAccept() send certificate to client but the HookState already set to DONE. the default certificate will be used and no Cert Hooks called. I will be update patch soon.

sorry ...

**Comment by bryancall** [comment]
@shinrich is going to shepard this.

**Pull request closed without merging**